proposal: llm contribution policy (#880)
jupyterhub/team-compass: docs/contribute/llm.md

(llm-policy)=

# Generative AI/LLM Contribution Policy

```{note}
"AI" herein refers to generative AI tools like large language models (LLMs)
that can generate, edit, and review software code, create and manipulate
images, or generate human-like communication.
```

We ask that all contributions (through issues and pull requests) are made by a human who takes responsibility for the code, documentation, or comment they submit.

> In short: LLM-generated contributions are not accepted,

Reviewer: I really like the language that people must take responsibility for their code.

> 
> ```{note}
> "AI" herein refers to generative AI tools like large language models (LLMs)
> that can generate, edit, and review software code, create and manipulate
> images, or generate human-like communication.
> ```
> 
> We ask that all contributions (through issues and pull requests) are made by a human who takes responsibility for the code, documentation, or comment they submit.
> In short: LLM-generated contributions are not accepted,
> and please make sure you are comfortable saying "I wrote this" about everything you submit to the project.
> 

Reviewer: ````suggestion
We ask that all contributions (through issues and pull requests) are made by a human who takes responsibility for the code, documentation, or comment they submit.

See [our LLM policy](#llm:policy) below. Here's a brief summary:

- **Responsibility**: You are responsible for any code you submit to JupyterHub's repositories, regardless of whether it was manually written or generated by AI.
- **Disclosure**: You must disclose whether AI has been used to assist in the development of your pull request.
- **Code Quality**. We will reject pull requests that we deem being [AI slop](https://en.wikipedia.org/wiki/AI_slop).
- **Copyright**. We reserve the right to reject any pull requests, AI generated or not, where the copyright is in question.
- **Communication**. When interacting with developers (forum, discussions, issues, pull requests, etc.) do not use AI to speak for you (except for translation).
- **AI Agents**. The use of an AI agent that writes code and then submits a pull request autonomously is not permitted.

```{note}
"AI" herein refers to generative AI tools like large language models (LLMs)
that can generate, edit, and review software code, create and manipulate
images, or generate human-like communication.
```
````

Reviewer: This replaces our overall summary with a brief, one sentence summary from each of the sub-sections in "policy". It just copy/pastes the most actionable/explicit statement from each section below (to my best judgment).

I think this makes the information more immediately glanceable w/ a link below for more details. I also moved this above the "note" so we have the most important information at the top of the document.

* **our global society**: we seek to minimise environmental impact and human exploitation in the development and deployment of JupyterHub infrastructure.

## Our concerns

"Generative AI" tools, such as LLMs, are

* contributing to rapid **over-burdening** of the reviewing capacity for many open source projects.
  * Unlike human contributors, models _cannot_ learn from feedback, vastly diminishing the long-term community benefit of the review process.
* trained on very large datasets, leaving many unsettled questions about copyright and consent. [US Copyright guidance](https://www.copyright.gov/ai/Copyright-and-Artificial-Intelligence-Part-2-Copyrightability-Report.pdf)

And in particular, the models in use by the commercial AI industry, which account for the vast majority of LLM-generated contributions today, are particularly destructive, as they are:

* trained on very large datasets without credit or consent and do not respect the licenses (or lack thereof) of this input data.
  * Even _if_ copyright of LLM output is settled safely worldwide, the training of models on open-licensed and/or proprietary inputs and generating outputs stripped of credit remains objectionable.

> * consuming very large amounts of energy and potable water, both during their training phases and as they are used to generate outputs. See the MIT Technology Review's summaries of the literature at their [Super Topic: AI and our energy future](https://www.technologyreview.com/supertopic/ai-energy-package/).
> 
> ## Policy
> 

Reviewer: ```suggestion

(llm:policy)=
## Policy

```

Reviewer: so we can link to the explicit policy part in links etc

### Responsibility

You are responsible for any code you submit to JupyterHub's repositories, regardless
of whether it was manually written or generated by AI. You must understand and be able
to explain the code you submit as well as the existing related code. It is not
acceptable to submit a patch that you cannot understand and explain yourself.
In explaining your contribution, do not use AI to automatically generate
descriptions. Always make sure you are comfortable saying "I wrote this"
before submitting code or a comment.

### Disclosure

You must disclose whether AI has been used to assist in the development of
your pull request.
If so, you must document which tool(s) have been used, how they were used,

> and specify what code or text is AI generated. We will reject any pull request

Reviewer: This feels to me like something that will be tough to consistently enforce in practice. Is every maintainer prepared to commit to adding an LLM disclosure statement to every single PR they make, no matter how large or small, no matter the circumstances?

If not then it might help to clarify the conditions under which this is required and when it is not (e.g., for example, if number of LOC changed is above N, if it's a new contributor, or something like an honor system where we ask people to just be honest about it if they did use an LLM, etc)

that does not include the disclosure.

> 

Reviewer: ```suggestion
If you've used AI in any part of the workflow that led to your contribution, you must disclose that you used AI and how you used it.
Document which tool(s) have been used, how they were used,
and specify what code or text is AI generated. We will reject any pull request
that does not include the disclosure.

```

Reviewer: I suspect this will be an easier policy to enforce. Otherwise we're asking every contributor to add a statement about AI and LLMs to every PR regardless of whether they used LLMs or not. That's setting us up to _block_ every PR that doesn't include this statement. I suspect that in practice, this policy would end up being unevenly enforced, so I softened the requirement a bit to avoid that. 

I don't hold this strongly though, so feel free to discard!

Author: I could go either way - I see both value and tedium in requiring "I did not use AI" on ~all PRs. Without it, the absence is a lack of information, whereas with it, the lack of a statement is immediately suspect. I think your suggestion is an easier, less confrontational place to start from, so think it's a good idea.

Reviewer: Yeah - I think that makes sense too. Maybe another way to resolve this is to add a "we'll use our best judgment" type of clause in there? Here are things I'm trying to avoid:

- A strict policy but that we just don't enforce for a subset of privileged people in under-specified ways.
- A strict policy that we regularly don't enforce because "it's obvious" (I'd rather we just add the judgment call language in there instead)
- A strict policy that adds unhelpful toil to an already over-burdened team (e.g. "make a PR fixing a one-line typo" -> "lose a day or two of back-and-forth because the author did not an AI tool use statement in the PR")

Personally I think it's OK to couch language like "we reserve the right to" to give maintainers the ability to make judgment calls.

Author: Yeah, let's start with your suggestion and go from there

Because closed commercial LLMs do not appropriately attribute their training set, their output is never acceptable.

### Communication

When interacting with developers (forum, discussions,
issues, pull requests, etc.) do not use AI to speak for you (except for
translation). If the developers want to chat with a chatbot,
they can do so themselves. Human-to-human communication is essential for an
open source community to thrive.

### AI Agents

The use of an AI agent that writes code and then submits a pull request autonomously is
not permitted.
We will not include or accept supporting resources in `AGENTS.md`, `CLAUDE.md` or similar files to our repos as we prefer human contributors follow the
[contribution guidelines](guide.md) that already exist.
If present, these files shall only include instructions to _prevent_ generating contributions, such as that [used by lobste.rs](https://github.com/lobsters/lobsters/blob/f2b1ccaa21af0721db73b77b7fb8bffe7ba12ffb/AGENTS.md).

### Other Resources

While these do not formally form part of JupyterHub's AI policy, the following resources
may be helpful in understanding some pitfalls associated with using AI to contribute to
JupyterHub:

> 

Reviewer: ```suggestion

## Other Resources

While these do not formally form part of JupyterHub's AI policy, the following resources
may be helpful in understanding some pitfalls associated with using AI to contribute to
JupyterHub:
```

Reviewer: just so it's clear this isn't part of the policy itself

(llm-policy)=

# Generative AI/LLM Contribution Policy

> We ask that all contributions (through issues and pull requests) are made by a human who takes responsibility for the code, documentation, or comment they submit.

Reviewer 2: ```suggestion
We ask that all contributions (through issues and pull requests) are made by a human who takes responsibility for the code, documentation, or comments they submit.
```

Minor style suggestion

See [our LLM policy](#llm:policy) below. Here's a brief summary:

- **Responsibility**: You are responsible for any code you submit to JupyterHub's repositories, regardless of whether it was manually written or generated by AI.
- **Disclosure**: You must disclose whether AI has been used to assist in the development of your pull request.

> - **Code Quality**. We will reject pull requests that we deem being [AI slop](https://en.wikipedia.org/wiki/AI_slop).

Reviewer 2: ```suggestion
- **Code Quality**. We will reject pull requests that we deem as [AI slop](https://en.wikipedia.org/wiki/AI_slop).
```

Minor style suggestion

> - **Copyright**. We reserve the right to reject any pull requests, AI generated or not, where the copyright is in question.

Reviewer 2: > Because closed commercial LLMs do not appropriately attribute their training set, their output is never acceptable.

I think there has to be mention of this hard-line at the top-level if contributors are coming here for a quick glance for how to get started with contributions. I echo @consideRatio 's sentiment that this gets lost in the details further down, and is important enough to highlight since this should have a material effect on a PR workflow and probably the most contentious effect of this policy.

Reviewer 3: > Because closed commercial LLMs do not appropriately attribute their training set, their output is never acceptable.

I fail to reconcile that statement with accepting any use of Claude Code, Codex, Copilot CLI, or autocomplete from associated companies' models that leads to any character written in the PR process.

I know almost nothing about copyright (especially not in an international context), but I figure with a statement like this retained in the policy, it would be better to clarify the statement implications concretely and discuss if we align and accept them rather than leaving it in for people to interpret differently. For me it reads extremely strict, while I don't pick up in this discussion that its meant to be interpreted that strict.

Reviewer 3: I made a suggestion to remove that sentence in https://github.com/jupyterhub/team-compass/pull/880#discussion_r3178493900

Reviewer 4: > while I don't pick up in this discussion that its meant to be interpreted that strict.

fwiw, the existence of this statement matches with my interpretation of this policy and the discussion here so far.

Reviewer 3: My perception is influenced mostly by https://github.com/jupyterhub/team-compass/pull/880#issuecomment-4306736361 and the fact that its not being stated at the top of the policy currently.

Reviewer 5: This definitely seems important enough to be at the top of the policy. But I also wouldn't complain if this was merged without that change, and would be happy to open a PR.

Reviewer 6: I would not interpret this statement @consideRatio as a ban on all AI tool usage. It definitely discourages the "throw it over the wall" and "lacking context of the project" AI slop. For contributors who may be unsure, they can ask or disclose what has been used as a first step which we can guide in a PR template.

I would recommend leaving as written for iteration 1. A separate issue can be created to discuss language improvement for a v2 of the policy.

@minrk I think this PR has reached general consensus. Let's aim for "good/well reasoned" for now over capturing everything. There's been wonderful input from the team across the board. I'm happy to press the green button to merge or leave it in your capable hands.

- **Communication**. When interacting with developers (forum, discussions, issues, pull requests, etc.) do not use AI to speak for you (except for translation).
- **AI Agents**. The use of an AI agent that writes code and then submits a pull request autonomously is not permitted.

## Principles

We, the JupyterHub community, value:

> * **human co-creation**: we are proud to collaborate with developers, researchers, educators, and infrastructure specialists across our global community.

Reviewer: ```suggestion
* **human co-creation**: we are proud to collaborate with developers, researchers, educators, and infrastructure specialists across our global community. We value contributors over their contributions.
```

One more quick thought, take it or leave it. I really liked this sentence from a recent [blog post from Simon W. about the Zig LLM policy](https://simonwillison.net/2026/Apr/30/zig-anti-ai/):

> Zig values contributors over their contributions.

I wonder if there's some way to work that into our values section. Here's a quick stab

(llm:policy)=
## Policy

### Responsibility

You are responsible for any code you submit to JupyterHub's repositories, regardless
of whether it was manually written or generated by AI. You must understand and be able
to explain the code you submit as well as the existing related code. It is not
acceptable to submit a patch that you cannot understand and explain yourself.
In explaining your contribution, do not use AI to automatically generate
descriptions. Always make sure you are comfortable saying "I am the author"
before submitting code or a comment.

### Disclosure

If you've used AI in any part of the workflow that led to your contribution, you must disclose that you used AI and how you used it.
Document which tool(s) have been used, how they were used,
and specify what code or text is AI generated. We will reject any pull request
that does not include the disclosure.

### Code Quality

Code generated by AI can be of low quality. Contributors are expected to
submit code that meets JupyterHub's standards. We will reject pull requests that we

> deem being [AI slop](https://en.wikipedia.org/wiki/AI_slop). Do not waste

Reviewer 2: ```suggestion
deem as [AI slop](https://en.wikipedia.org/wiki/AI_slop). Do not waste
```

Minor style suggestions

developers' time by submitting code that is fully or mostly generated by AI, and
doesn't meet our standards.

### Copyright

All code in JupyterHub is released under the BSD 3-clause copyright license.
Contributors to JupyterHub license their code under the same license when it is
included into JupyterHub's version control repository. That means contributors must
own the copyright of any code submitted to JupyterHub or must include the BSD
3-clause compatible open source license(s) associated with the submitted code
in the patch. Code generated by AI may infringe on copyright and it is the
submitter's responsibility to not infringe. We reserve the right to reject any pull
requests, AI generated or not, where the copyright is in question.

> Because closed commercial LLMs do not appropriately attribute their training set, their output is never acceptable.

Reviewer 3: I used `codex` with GPT 5.5 to generate a `kubectl patch` statement with a brief prompt asking me to patch a specific thing in a k8s Deployment. I disclaimed it in https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/3883. Should it be accepted or not?

> Because closed commercial LLMs do not appropriately attribute their training set, their output is never acceptable.

I don't want to violate our policy, but I read it as I do if we retain this part and its very discouraging. This PR suggestion removes it.

```suggestion
in the patch. Code generated by AI may infringe on copyright and it is the
submitter's responsibility to not infringe. We reserve the right to reject any pull
requests, AI generated or not, where the copyright is in question.
```

<details><summary>The `kubectl patch` command generated</summary>

```
kubectl patch deployment autohttps --type='strategic' -p='{"spec":{"template":{"spec":{"containers": [{"name":"traefik","command":["sh","-c","sleep 10 && /entrypoint.sh traefik"]}]}}}}'
```

</details>

Reviewer 3: See also related discussion in https://github.com/jupyterhub/team-compass/pull/880#discussion_r3178426908

Reviewer 6: There is probably a distinct difference between submitting the patch as generated verbatim (which draws a fine line of uncertainty in copyright infringement). If the same generated patch was refined to be human-readable versus one long line, I think that would be acceptable from my vantage point.

```
kubectl patch deployment autohttps --type='strategic' -p='{"spec":{"template":{"spec":{"containers": [{"name":"traefik","command":["sh","-c","sleep 10 && /entrypoint.sh traefik"]}]}}}}'
```

becomes
```
kubectl patch deployment autohttps \
  --type='strategic' \
  -p='{
    "spec": {
      "template": {
        "spec": {
          "containers": [
            {
              "name": "traefik",
              "command": [
                "sh",
                "-c",
                "sleep 10 && /entrypoint.sh traefik"
              ]
            }
          ]
        }
      }
    }
  }'
```

making it more readable as a submission and allows a reviewer to better determine if this is general knowledge vs. copyrighted information.

### Communication

When interacting with developers (forum, discussions,
issues, pull requests, etc.) do not use AI to speak for you (except for
translation). If the developers want to chat with a chatbot,
they can do so themselves. Human-to-human communication is essential for an
open source community to thrive.

### AI Agents

The use of an AI agent that writes code and then submits a pull request autonomously is
not permitted.
We will not include or accept supporting resources in `AGENTS.md`, `CLAUDE.md` or similar files to our repos as we prefer human contributors follow the
[contribution guidelines](guide.md) that already exist.
If present, these files shall only include instructions to _prevent_ generating contributions, such as that [used by lobste.rs](https://github.com/lobsters/lobsters/blob/f2b1ccaa21af0721db73b77b7fb8bffe7ba12ffb/AGENTS.md).

### Other Resources

While these do not formally form part of JupyterHub's AI policy, the following resources
may be helpful in understanding some pitfalls associated with using AI to contribute to

> JupyterHub:

Reviewer 2: I think there's some duplication here
```suggestion
```

# Generative AI/LLM Contribution Policy

We ask that all contributions (through issues and pull requests) are made by a human who is willing to take accountability for the code, documentation or comment they submit.

## Principles

We, the JupyterHub community, value:

* **human co-creation**: we are proud to collaborate with developers, researchers, educators, and infrastructure specialists across our global community.
  * We respect **their copyright** over their work and appreciate their shared investment in the scientific open source ecosystem by licensing the materials under the [BSD 3-Clause "New" or "Revised" License](https://github.com/jupyterhub/jupyterhub/blob/main/LICENSE).
  * We respect the **time** taken to read and review contributions, maintinaing a high standard and supportive community engagement.
* **security**: JupyterHub is trusted infrastructure for hundreds of thousands of users and we prioritize keeping their data, code, and personal configuration information secure.
* **veracity**: beyond security, we ensure our tools do what we think they do, and that we respect accuracy in communications within and beyond the scientific open source ecosystem.

> * **our global society**: we seek to minimise environmental impact and human exploitation in the development and deployment of JupyterHub infrastructure.

Reviewer 7: I'm glad this section resonates! I think its important to remind us all what we are FOR rather than just what we are against 💖 

I also didn't figure out how to include the political harms / threats to democracy .... but I'm very happy to include those (as they're one of the things that I am most concerned about!) Happy to iterate on any suggested phrasings (and I'll noodle myself) if folks think that would be a good additional bullet point?

> * **our global society**: we seek to minimise environmental impact and human exploitation in the development and deployment of JupyterHub infrastructure.

Reviewer: I like this section too! +1 to rooting ourselves in principles, values, and goals.

## Our concerns

Large language models (LLMs) are:

> * changing, and at time of writing (February 2026) **over-burdening**, the reviewing capacity for many open source projects. [REF]

Reviewer 5: Super helpful comment from me... I agree with both of you! I'm hopeful that this can get better _and_ I don't think that's useful information for the policy :) When creativity eventually prevails, we can update the policy with new information!

* trained on very large datasets, and most models do not respect the licenses (or lack thereof) of this input data. [REF]

>     * We are keen to follow the development of soverign models such as ["Apertus"](https://www.swiss-ai.org/apertus) which have been trained on appropriately copywrited data, fully transparently, and published with open weights.

Reviewer 7: Happy to soften this point about Apertus! 

I'm still excited about a sovereign and small model future.... but that's maybe a separate discussion than just tempering the point about Apertus ✔️

* in many cases trained by people who are underpaid and exposed to traumatic content without appropriate support.
  See Wired's article from 2024: [The Low-Paid Humans Behind AI’s Smarts Ask Biden to Free Them From "Modern Day Slavery"](https://www.wired.com/story/low-paid-humans-ai-biden-modern-day-slavery)
* consuming very large amounts of energy and potable water, both during their training phases and as they being used to generate outputs. See the MIT Technology Review's summaries of the literature at their [Super Topic: AI and our energy future](https://www.technologyreview.com/supertopic/ai-energy-package/).

## Policy

1. Only submit code or documentation to JupyterHub that you wrote and that you understand.
    * We have chosen this requirement given our concerns around **copyright**, **reviewer burden**, and to maintain **auditable accountability** for the **veracity** of our work.

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Reviewer 7: @minrk - can you explain why this phrasing is so important to you? I can imagine a few different reasons and I think it would be helpful to know the specifics.

The reason I softened the wording (and I did know I was softening it - thank you for engaging to suggest it back!) is that I have been in some conversations that imply that using a tool (any tool!) but then writing out the submission by hand is ok.... and that feels like such an awkward recommendation!

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Reviewer 5: The wording "I wrote this" doesn't resonate with me, it feels more about the mechanics of who/what typed the code than who is intellectually responsible for it. I don't know how to do better right now but I am also interesting in hearing an argument for this wording :)

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Author: I appreciate that! FWIW, that's why I chose "wrote" as opposed to "typed". Trying to get to a higher-level evaluation of authorship, and "wrote" was the. best word I could think of for that. I particularly want to be able to distinguish between an ownership sense between:

1. "I wrote this, and used some tools to help me do it" (OK), vs
2. "I got Claude to write this, and I reviewed it and vouch for it" (not OK)

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Reviewer 5: "I authored this"? Does that resonate with you? To me it feels closer to what you're trying to communicate. And the longer version of "I authored this" might be "I did the thinking, and I can explain every decision".

Even "I got Claude to write this, and I reviewed it and vouch for it" (not OK) doesn't quite resonate with me -- I did something like this recently that I think would have fallen in to this "not OK" category based on my current interpretation. I had a large configuration that was written in JSON and was very confusing (expressed as two files with unclear boundaries), and it needed a more expressive medium and clear design. I defined a data model and had Claude migrate the config to Python objects, and it did a great job, I just reviewed it and made minor adjustments. It's a very rote activity that I would not have wanted to "write" myself (and I have an RSI, extra motivation to not write it myself), but I would have produced the same outcome if I had "written" it. But again, this may come down to we mean slightly different things by "write".

Full disclosure, I couldn't find a better word than "wrote" and Claude helped me out :laughing: "authored" feels less mechanical to me than "wrote", which feels less mechanical than "typed".

I feel guilty that this feels a bit nitpicky. But I hope it's useful :upside_down_face:

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Author: To me,  "authored" doesn't mean anything different from "wrote" other than using a noun as a verb (which I personally really dislike style-wise, sorry for being nitpicky myself!). We already have a verb for what authors do: "write", which to me is distinct from the _mechanical_ operation. But I agree with the goal you express. Maybe we just can't have a version that's both short and clear.

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Reviewer 5: Interesting! I've always thought of "to author" as a valid verb for "to create". I did a bit of googling since I felt challenged by your comment and found, and the first result was bizarrely relevant to exactly this conversation :laughing:  https://www.merriam-webster.com/grammar/author-as-a-verb

>Maybe we just can't have a version that's both short and clear.

I sadly agree this is often true. But I would _love to_. Please let me know if the article I linked changes your thoughts at all!

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Author: Thanks for the link! Yeah, I definitely don't think it's _wrong_, the phrase just feels off to me. Genuinely a personal preference for me. I just also feel like "I wrote this" also means "I created this," but I also ackmowledge a policy is for _resolving_ ambiguity, so if folks read it differently, that's important to resolve!

I'd be happier with "I created this" or "I made this" than "I authored this"

<img width="513" height="461" alt="imadethis" src="https://github.com/user-attachments/assets/bc7d09c8-334d-4172-a9a4-261c700e9719" />

>     * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of LLMs for supporting their open source contributions.

Reviewer 5: I like those too. And I was debating whether to attach that exact image yesterday :laughing:

  * We respect the **time** taken to read and review contributions, maintaining a high standard and supportive community engagement.
* **security**: JupyterHub is trusted infrastructure for hundreds of thousands of users and we prioritize keeping their data, code, and personal configuration information secure.
* **veracity**: beyond security, we ensure our tools do what we think they do, and that we respect accuracy in communications within and beyond the scientific open source ecosystem.
* **our global society**: we seek to minimise environmental impact and human exploitation in the development and deployment of JupyterHub infrastructure.

## Our concerns

Large language models (LLMs), the primary underlying technology used in "Generative AI" products, are:

* contributing to rapid **over-burdening** of the reviewing capacity for many open source projects. [REF]
* trained on very large datasets, and all widely used models do not respect the licenses (or lack thereof) of this input data. [REF]
* in many cases trained by people who are underpaid and exposed to traumatic content without appropriate support.

>   See Wired's article from 2024: [The Low-Paid Humans Behind AI’s Smarts Ask Biden to Free Them From "Modern Day Slavery"](https://www.wired.com/story/low-paid-humans-ai-biden-modern-day-slavery)

Author: Energy harms would be unchanged, output failing to appropriately attribute and credit derivative works would remain, slop output would be unchanged, unlicenseable output is unchanged. So the _only_ cases where there's any variation are the ones which are already qualified with "in many cases" which I think is overstated in terms of models (if not usage), and "all widely used models" which is also factual. I don't see where we aren't drawing a factual line.

Could you propose some specific small tweaks you're asking for?

* consuming very large amounts of energy and potable water, both during their training phases and as they being used to generate outputs. See the MIT Technology Review's summaries of the literature at their [Super Topic: AI and our energy future](https://www.technologyreview.com/supertopic/ai-energy-package/).

> * generating output that may not be able to be covered by our licenses and copyright, at least in the United States. [US Copyright guidance](https://www.copyright.gov/ai/Copyright-and-Artificial-Intelligence-Part-2-Copyrightability-Report.pdf)

Reviewer 8: I think this a re-framing of the point above, maybe we can combine them?  It is a good point that "all LLMs" (including open LLMs" are trained on open source software.  I do think there is an open question -- at least US courts have not established the answer in the US legal system -- as to whether that makes them a 'derivative product'.  I don't think this case was imagined by copyright law before now, reasonable people could disagree.  

But I do think there is a material difference between an model like Olmo that says: 'here are exactly the the data we trained on', so at least anyone can inspect and see if their code is included, and commercial models that don't come anywhere close to this disclosure.  As an open source community, it seems odd for us to even nod to that distinction.  

I agree it would be perfectly reasonable to say "we don't accept contributions from any LLM due to open questions about the copyright and IP of the content they create".

  * We respect the **time** taken to read and review contributions, maintaining a high standard and supportive community engagement.
* **security**: JupyterHub is trusted infrastructure for hundreds of thousands of users and we prioritize keeping their data, code, and personal configuration information secure.
* **veracity**: beyond security, we ensure our tools do what we think they do, and that we respect accuracy in communications within and beyond the scientific open source ecosystem.
* **our global society**: we seek to minimise environmental impact and human exploitation in the development and deployment of JupyterHub infrastructure.

## Our concerns

Large language models (LLMs) are:

* contributing to rapid **over-burdening** of the reviewing capacity for many open source projects. [REF]
* trained on very large datasets, and all widely used models do not respect the licenses (or lack thereof) of this input data. [REF]
    * We are keen to follow the development of sovereign models such as ["Apertus"](https://www.swiss-ai.org/apertus) which have been trained fully transparently, and published with open weights.

>       Though even these generally fail to respect the license and copyright of the works their output derives from.

Reviewer 4: ```suggestion
```

I would just remove this for now, or move this to an entirely different section, as it's two contradictory sentences that cancel each other out.

* consuming very large amounts of energy and potable water, both during their training phases and as they being used to generate outputs. See the MIT Technology Review's summaries of the literature at their [Super Topic: AI and our energy future](https://www.technologyreview.com/supertopic/ai-energy-package/).
* generating output that may not be able to be covered by our licenses and copyright, at least in the United States. [US Copyright guidance](https://www.copyright.gov/ai/Copyright-and-Artificial-Intelligence-Part-2-Copyrightability-Report.pdf)

## Policy

1. Only submit code or documentation to JupyterHub that you wrote and that you understand.
    * We have chosen this requirement given our concerns around **copyright**, **reviewer burden**, and to maintain **auditable accountability** for the **veracity** of our work.
    * We leave the interpretation of this request to you: tool-assisted coding and automation leaves some gray areas and we recognize that everyone has different perspectives on the use of various tools for supporting their open source contributions.

>     * Always make sure you are comfortable saying "I wrote this" before submitting code or a comment.

Reviewer 4: I like this a lot, and I would suggest elevating this to the 'in short' section as well.

    * We strongly endorse using deterministic tools to check and improve the quality of your code, such as automated code dependency updates, linter auto-fixes, refactor tools, etc.
2. Communicate in issue and pull request threads in your own words.
  Please do not use the verbatim output of an LLM in conversation with our user, reviewer and maintainer community.
    * We have chosen this requirement given our priority to **value human co-creation**.
3. Will will not include or accept supporting resources in `AGENTS.md`, `CLAUDE.md` or similar files to our repos as we prefer human contributors follow the [contribution guidelines](guide.md) that already exist.

>    If present, these files shall only include instructions to _prevent_ generating contributions.

Reviewer 4: I like this
